### PR TITLE
Network ID logic should not be applied to pseudo-transactions

### PR DESCRIPTION
**Pull request by dangell7**
This PR fixes the issue with enabling new amendments on a network with an ID > 1024. 

See an example here: https://github.com/XRPLF/rippled/issues/4736

- [Context on pseudo-transactions](https://xrpl.org/pseudo-transaction-types.html)

### CODE REVIEW

**Comment by intelliot** [comment]
Just curious - is there a reason this solution is better than adding the NetworkID field to pseudo-transactions?

**Comment by mvadari** [comment]
> Just curious - is there a reason this solution is better than adding the NetworkID field to pseudo-transactions?

I don't fully understand how pseudo-transactions work, but is there a chance that a signed pseudo-transaction could be submitted on another network to cause unwanted changes?

**Comment by ximinez** [comment]
Pseudo-transactions don't get signed, nor are they submitted. I think it's possible to request one, but I'm not 100% sure of that. Point is, that's not a risk.

That said, adding the NetworkID to pseudo-transactions _should_ be an equally viable solution. Perhaps better, because someone seeing one of these pseudo-txs would know exactly where it came from.

**Comment by RichardAH** [comment]
> That said, adding the NetworkID to pseudo-transactions _should_ be an equally viable solution. Perhaps better, because someone seeing one of these pseudo-txs would know exactly where it came from.
> 
> P.S. Either way, unit tests should be added to ensure that pseudo-transactions are generated and processed correctly with a network ID > 1024.

It's not obvious to me why pseudos would ever need a network ID? If you want to differentiate networks just add NetworkID to the fees (or another singleton) object?

**Comment by intelliot** [comment]
It would never be _needed_. When looking some pseudo-transaction JSON, you would be able to see which network it came from, however.

File: src/ripple/app/tx/impl/Transactor.cpp [hunk 1]
```diff
@@ -40,25 +40,32 @@ namespace ripple {
 NotTEC
 preflight0(PreflightContext const& ctx)
 {
-    uint32_t const nodeNID = ctx.app.config().NETWORK_ID;
-    std::optional<uint32_t> const txNID = ctx.tx[~sfNetworkID];
-
-    if (nodeNID <= 1024)
+    if (isPseudoTx(ctx.tx))
```
Comment (gregtatcam) [inline src/ripple/app/tx/impl/Transactor.cpp:43]: It's better not to have the empty block and check `if (!isPseudoTx(ctx.tx))`

Comment (gregtatcam) [reply src/ripple/app/tx/impl/Transactor.cpp:43]: I'm not having much luck with testing this fix. I run a local network of 5 nodes. I added this to config:
>      [network_id]
>      1026
>      [amendment_majority_time]
>      16 minutes
I ran this to every node:
>              {
>                  "method": "feature",
>                  "params": [
>                    {
>                      "feature": "AMM",
>                      "vetoed": false
>                    }
>                  ]
>                }
This is the response:
>       {'count': 1,
>         'enabled': False,
>         'name': 'AMM',
>         'supported': True,
>         'threshold': 3,
>         'validations': 5,
>         'vetoed': False}

Comment (gregtatcam) [reply src/ripple/app/tx/impl/Transactor.cpp:43]: I'll re-test with the suggested wait time to double check.

Comment (gregtatcam) [reply src/ripple/app/tx/impl/Transactor.cpp:43]: I waited 45 minutes - the same result.

**Comment by intelliot** [comment]
Notes
- Concern: consistency, and having all transactions follow the same rules
- May be preferable to avoid making exceptions which may be forgotten in the future
- Should try to minimize exceptions that apply only to pseudo-transactions

**Comment by intelliot** [comment]
Just to close the loop on my earlier question: I'm personally fine with this solution, given these are pseudo-transactions, which are already different from normal transactions anyway. It makes sense to review (and potentially merge) this change as-is.

File: src/ripple/app/tx/impl/Transactor.cpp [hunk 2]
```diff
@@ -40,25 +40,33 @@ namespace ripple {
...
     {
-        // legacy networks have IDs 1024 and below. These networks cannot
-        // specify NetworkID in txn
-        if (txNID)
-            return telNETWORK_ID_MAKES_TX_NON_CANONICAL;
+        // all emitted and pseudo transactions are free to pass, do not need
+        // network id
```
Comment (seelabs) [inline src/ripple/app/tx/impl/Transactor.cpp:46]: What should we do if a pseudo transaction has a network id? Right now it's ignored, which is probably fine. But some other options would be:
1) If present, it needs to follow the same rules as other transactions (i.e. match the configured network id), fail if the network is less than 1024.
2) Fail if present in any form.
3) Assert if present in any form, but allow to continue processing in release mode.

I'd probably prefer (3). If someone does add a network id to a pseudo transaction, and they did it intentionally, the assert will let them know this code needs to change. But you're call if you want to add it or not. I'm fine either way.

As a side note, I also prefer not to have the empty `if` block, but that's just a preference. I'm also fine if you want to keep it as-is.

Comment (ximinez) [reply src/ripple/app/tx/impl/Transactor.cpp:46]: I'd vote for option 1. It has the advantage of being simple to add that check. Simply change the check to
```
if (isPseudoTx(ctx.tx) && !ctx.tx.isFieldPresent(sfNetworkID))
```

Comment (dangell7) [reply src/ripple/app/tx/impl/Transactor.cpp:46]: I like option 1 because its simple.

Comment (seelabs) [reply src/ripple/app/tx/impl/Transactor.cpp]: @dangell7 I don't think that patch is going to work If the network ID is required, but isn't present the new patch let's the transaction succeed.

Comment (dangell7) [reply src/ripple/app/tx/impl/Transactor.cpp]: fixed in [d8b2ba4](https://github.com/XRPLF/rippled/pull/4737/commits/d8b2ba42a6b2ba1a327ecb7f8021bdff662f3e75)

**Review by seelabs — APPROVED** [review]
:+1:

**Review by gregtatcam — APPROVED** [review]
👍

**Comment by intelliot** [comment]
note: needs to be brought up-to-date with `develop`

**Comment by sgramkumar** [comment]
Verified on networks with network_id 12 & 1212. Amendment is enabled in both cases.

**Pull request merged**